class.c

```diff
@@ -371,6 +376,79 @@ rb_mod_init_copy(VALUE clone, VALUE orig)
 	rb_id_table_foreach(RCLASS_M_TBL(orig), clone_method_i, &arg);
     }
 
+    if (RCLASS_ORIGIN(orig) == orig) {
+        RCLASS_SET_SUPER(clone, RCLASS_SUPER(orig));
+    }
+    else {
+        VALUE p = RCLASS_SUPER(orig); 
+        VALUE orig_origin = RCLASS_ORIGIN(orig);
+        VALUE prev_clone_p = 0;
+        VALUE origin_stack = rb_ary_tmp_new(2);
+        VALUE origin[2];
+        VALUE clone_p = 0;
+        long origin_len;
+        int add_subclass;
+        VALUE clone_origin;
+
+        rb_ensure_origin(clone);
+        clone_origin = RCLASS_ORIGIN(clone);
+
+        while (p && p != orig_origin) {
+            if (BUILTIN_TYPE(p) != T_ICLASS) {
+                rb_bug("non iclass between module/class and origin");
+            }
+            clone_p = class_alloc(RBASIC(p)->flags, RBASIC(p)->klass);
+            if (prev_clone_p) {
```

Comment:
This check would not be necessary if `prev_clone_p` is initialized to `clone`, unless I miss something.

Reply:
I agree. I'll make this change.